drake/solvers/Optimization.h

```diff
@@ -246,20 +278,27 @@ namespace Drake {
 //    const DecisionVariable& addIntegerVariables(size_t num_new_vars, std::string name);
 //  ...
 
-    void addCost(const std::shared_ptr<Constraint>& obj) {
-      checkVariables(obj);
+    void addCost(std::shared_ptr<Constraint> const& obj, VariableList const& vars) {
```

Comment:
i think we're on the same page, but I would like to (soon) support some reflection here... so that if you call addCost (or addConstraint) with a Constraint that has more structure than the generic structure, then it will dispatch to e.g. addLinearConstraint.